### DISCUSSION TITLE
ROSCO and Locally-compiled OpenFAST

### DISCUSSION THREAD

**Discussion by Yuksel-Rudy** [NatLabRockies/ROSCO#151] [Q&A] [answered]
Hello,
I have ROSCO fully installed in a conda environment and I'm trying to run example 06 but with a recompiled OpenFAST. Instead of using the command `conda install -c conda-forge openfast` which generates the executable for the current version of OpenFAST, what can I do to point ROSCO to using a custom recompiled OpenFAST file? I tried installing OpenFAST first with the above command then copy/paste all the files in the bin folder in OpenFAST/install to the bin folder in my ROSCO environment and the example ran successfully but the CPU time for the example went from 5.5 seconds with the normal installation to 2 minutes. So I think there's a compatibility issue because the size of OpenFAST file with the default setup has a different size compared to the recompiled version. (Just to be clear, I haven't really did any changes in the OpenFAST src code just yet, I just basically recompiled from source). I've noticed that anaconda also created a package outside the environment in anaconda3\pkgs\openfast-3.1.0-hdde1195_1 folder. I tried replacing the bin there, too but it's still slow. What am I doing wrong? 

P.S. I'm using WSL, Ubuntu 20.04.4

**Comment by rafmudaf** [comment, marked as answer]
Hi @Yuksel-Rudy It sounds like you successfully got ROSCO to use your locally-compiled OpenFAST, so I think the question in the title of this discussion is not the issue. Is that correct?

As for the performance and size of OpenFAST, how did you compile it? What CMake settings did you use? There are a few settings that will affect the performance and size of the file, among other things, so you have to be careful to set it up correctly. If you want to match the conda configuration, see https://github.com/conda-forge/openfast-feedstock/blob/main/recipe/build.sh#L9.

Additionally, you'll get better performance using the Intel Fortran compiler and MKL library.

**Reply by Yuksel-Rudy** [reply]
@rafmudaf Thank you for your quick response.

I executed CMake with the default options: `cmake..`
then compiled everything with the command `make` and moved binaries to openfast/install with `make install`

I tried your option of cmake -DDOUBLE_PRECISION=OFF after making build dir, but that gave the following error:
CMake Error: The source directory "/home/rudy/OpenFAST_clean/OpenFAST/build" does not appear to contain CMakeLists.txt.
Specify --help for usage, or press the help button on the CMake GUI.

Can you elaborate more on how to mimic conda configuration?

Thank you

**Reply by rafmudaf** [reply]
Since you're going to be compiling, I recommend reading through the [Understanding CMake](https://openfast.readthedocs.io/en/main/source/install/index.html#understanding-cmake) section of the OpenFAST documentation. Also consider reading the linked article there from CMake on [Running CMake](https://cmake.org/runningcmake/).

In your case, the command you wrote here does not include the `..`. You have to create the build directory and then move into it. From there, run CMake and point to one directory up:

```bash
mkdir build
cd build
cmake .. -DDOUBLE_PRECISION=OFF
```

The conda build file uses a different form to point to the source directory. There, the `-S ${SRC_DIR}` part is doing the equivalent of `..` for you.

**Reply by Yuksel-Rudy** [reply]
Thank you @rafmudaf, that actually helped with the performance issue.

**Reply by rafmudaf** [reply]
Yea, single precision is quite a bit faster, but the computations are less precise for most modules. In most cases, that is totally fine and worth the speed up.
